packages/agent-utils/src/agent/basicFunctionCallLoop.ts

```diff
@@ -23,21 +27,32 @@ export async function* basicFunctionCallLoop<TContext extends { llm: LlmApi, cha
 
     if (response.function_call) {
       const { name, arguments: args } = response.function_call;
-      const result = await executeAgentFunction(name, args, context, agentFunctions);
+      const { result, functionCalled } = await executeAgentFunction(name, args, context, agentFunctions);
+
+      let output: StepOutput;
 
       if (result.ok) {
         chat.temporary({ role: "system", name, content: result.value.content});
-        yield StepOutput.message(result.value);
-      }
-      else {
+        output = StepOutput.message(result.value);
+      } else {
         chat.temporary("system", result.error as string);
-
-        yield StepOutput.message({
+        output = StepOutput.message({
           type: "error",
           title: `Failed to execute ${name}!`,
           content: result.error as string
         });
-      } 
+      }
+
+      const terminate = functionCalled && shouldTerminate(functionCalled);
```

Comment:
Should we terminate even if the function call failed?

Reply:
If `functionCalled` is defined, that means we've called the function. In the case of `onGoalAchieved` this works fine.

Is there a case where we might call a termination function but it fails in some way so we want to continue looping? I think the `writeFunction` terminating function in the js-script-writer is the same scenario.

Reply:
At the moment no, but if we extend any of the terminating functions, e.g. we add a message to onGoalFailed, and if the agent gives wrong args e.g. an object instead of a string for the message, then we would terminate without letting the agent recover from his mistake

Reply:
Ah okay, so to solve this I think we can pass the function's result to the `should Terminate` callback, so the logic can be customized to check for this.